<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Find and filter Framework Initiative data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!DOCTYPE html>

<div class="sidenav">
  <a href="index.html">Home</a>
  <a href="registration_login.html">Register and log-in</a>
  <a href="find_filter_download.html">Find and filter</a>
  <a href="programmatic_access.html">Download data and meta-data</a>
  <a href="metadata_filter_ausmicrobiome.html">How to filter Australian Microbiome data</a>
</div>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Find and filter Framework Initiative data</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#search-box">Search box</a></li>
<li><a href="#meta-data-fields-available-for-search">Meta-data fields available for search</a></li>
<li><a href="#simple-search">Simple search</a></li>
<li><a href="#advanced-search-for-specific-meta-data-fields">Advanced search for specific meta-data fields</a></li>
<li><a href="#advanced-search-across-multiple-meta-data-fields">Advanced search across multiple meta-data fields</a></li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
</ul>
</div>

<style type="text/css">
.main-container {
  max-width: 1500px;
}
</style>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>The instructions here provide clarity regarding how to effectively search the Bioplatforms data portal, and obtain search results that are either specific to a certain meta-data field (e.g. <code>flowcell_id</code>), or which can capture the existence of search values across meta-data fields.</p>
<p>When you have found the data set(s) you were looking for, it’s time to <a href="programmatic_access.md">download</a>.</p>
</div>
<div id="search-box" class="section level1">
<h1>Search box</h1>
<p>The query terms for any search should be entered into the “Search” box, which appears on the landing pages for each Framework Initiative on the data portal.</p>
<p align="center">
<img src="images/search_1.png">
</p>
<p>The proper syntax for a data portal search is <code>[meta-data field name]:[value]</code>. Examples are provided in each of the sections below.</p>
<p><strong>Note</strong>: there are two search modes that can be used for the data portal.</p>
<ol style="list-style-type: decimal">
<li><strong>Simple</strong>: no colon is used, simply enter the search words</li>
<li><strong>Advanced</strong>: at least one colon present, with numerous options possible including boolean operators (<code>AND</code> / <code>OR</code> / <code>NOT</code> / <code>+</code> / <code>-</code> )</li>
</ol>
</div>
<div id="meta-data-fields-available-for-search" class="section level1">
<h1>Meta-data fields available for search</h1>
<p>For the Oz Mammals Genomics example above, some common meta-data fields that may be of interest are:</p>
<ul>
<li><code>bpa_dataset_id</code>: Bioplatforms Australia dataset ID for this sequencing run</li>
<li><code>bpa_library_id</code>: Bioplatforms Australia library ID for this library prep</li>
<li><code>bpa_sample_id</code>: unique identifier from Bioplatforms Australia, assigned to the material sample</li>
<li><code>class</code>: the full scientific name of the class in which the taxon is classified</li>
<li><code>common_name</code>: common name of the species</li>
<li><code>data_type</code></li>
<li><code>family</code>: the full scientific name of the family in which the taxon is classified</li>
<li><code>order</code>: the full scientific name of the order in which the taxon is classified</li>
<li><code>phylum</code>: the full scientific name of the phylum in which the taxon is classified</li>
<li><code>genus</code>: the full scientific name of the genus in which the taxon is classified</li>
<li><code>species</code>: the name of the first or species epithet of the scientific name</li>
<li><code>subspecies</code>: if relevant, the name of the lowest or terminal infraspecific epithet of the scientific name, excluding any rank designation, e.g. subspecies</li>
<li><code>library_type</code>: type of library prepared (exon, capture, ddRAD etc.)</li>
<li><code>life_stage</code>: the age class or life stage of the animal at the time of sampling</li>
<li><code>omg_project</code>: name of OMG project this work relates to (e.g. Phylogenomics)</li>
<li><code>sequencing_facility</code>: name of the sequencing facility</li>
<li><code>sequencing_platform</code>: name of sequencing platform used</li>
<li><code>sex</code>: the sex of the animal</li>
<li><code>tissue_type</code>: the type of tissue or substance sampled</li>
</ul>
<p><strong>Note</strong>: many more fields are in active use at any one time and these will vary slightly for each Framework Initiative. You can find a list of these fields by clicking on a relevant dataset and scrolling down to the ‘Additional Info’ table.</p>
</div>
<div id="simple-search" class="section level1">
<h1>Simple search</h1>
<p>To conduct a broad search, simply type a keyword into the search bar. This type of search will look for that keyword anywhere in the database and is more suited for initial discovery searches to gain an appreciation for the resources available through any one Framework Initiative.</p>
<p>For example, if you search the OMG section of the portal for data sets using the keyword “Dunnart”, the following occurs:</p>
<ul>
<li>33 data sets were discoverable (2020-10-14)</li>
<li>These are split across multiple <code>data_types</code> and <code>omg_project</code> values</li>
</ul>
<p align="center">
<img src="images/search_2.png">
</p>
</div>
<div id="advanced-search-for-specific-meta-data-fields" class="section level1">
<h1>Advanced search for specific meta-data fields</h1>
<p>If the values of interest are known, it is better to search for specific values (e.g. “Fat-tailed Dunnart”) in a single defined meta-data field (e.g. <code>common_name</code>).</p>
<p>The format for this is <code>field:[”value”]</code></p>
<p>For example, if you wanted to search for all data sets that were produced using a specific <code>sequencing_platform</code> making use of the accepted Bioplatforms sequencing platform ontology (e.g. searching for “PacBio Sequel”), your search syntax would look like this:</p>
<pre><code>sequencing_platform:&quot;PacBio Sequel&quot;</code></pre>
<p><strong>Note</strong>: make sure you include the apostrophes!</p>
<p>Here, 4 data sets are located, each of which used the PacBio Sequel instrument as the <code>sequencing_platform</code>.</p>
<p align="center">
<img src="images/search_3.png">
</p>
</div>
<div id="advanced-search-across-multiple-meta-data-fields" class="section level1">
<h1>Advanced search across multiple meta-data fields</h1>
<p>If you wanted to search across multiple meta-data fields, and each search term was mandatory / required, the search syntax would look like this:</p>
<pre><code>[meta-data field name]:[value] AND [meta-data field name]:[value]</code></pre>
<p>For example, to identify a specific combination of genus and species:</p>
<pre><code>genus:Sminthopsis AND species:crassicaudata</code></pre>
<p>This is useful as it distinguishes datasets for the genus Sminthopsis, for which there are 5 species available, including:</p>
<ul>
<li><em>Sminthopsis dolichura</em>,</li>
<li><em>Sminthopsis gilberti</em>,</li>
<li><em>Sminthopsis granulipes</em>,</li>
<li><em>Sminthopsis longicaudata</em>, in addition to</li>
<li><em>Sminthopsis crassicaudata</em></li>
</ul>
</div>
<div id="acknowledgements" class="section level1">
<h1>Acknowledgements</h1>
<p>Source material for <a href="https://gist.github.com/davidmiller/f22c2dcb347f3dbc4b16cb8a4bddbc04">CKAN search instructions</a></p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
