<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Programmatic access to the Bioplatforms Data Portal</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Home</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="registration_login.html">Registration &amp; login</a>
</li>
<li>
  <a href="find_filter_download.html">Find, filter &amp; download</a>
</li>
<li>
  <a href="programmatic_access.html">Programmatic access</a>
</li>
<li>
  <a href="metadata_filter_ausmicrobiome.html">Filter AusMicrobiome data</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Programmatic access to the Bioplatforms Data Portal</h1>

</div>


<p>The Bioplatforms Data Portal is based upon <a href="https://ckan.org/">CKAN</a>, an open source platform for sharing metadata and data. CKAN provides Application Programming Interfaces (APIs) which allow computer software to interact with the portal. It is possible for software to programmatically do everything you can do manually by using the portal in your web browser - including searching for data, downloading data, and even uploading data.</p>
<p>This guide covers access to the portal using the R and Python programming languages, and via the command shell on your desktop computer (<code>bash</code> or <code>zsh</code>.)</p>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>When your computer program connects to the portal, it must identify itself. It will do this by sending an ‘API key’. You can find this key on the portal.</p>
<ol style="list-style-type: decimal">
<li>Go to the portal <a href="https://data.bioplatforms.com/" class="uri">https://data.bioplatforms.com/</a></li>
<li>At the top right of the page, you will see either a prompt to log in, or your name</li>
<li>Log in, if required, and then click on your name</li>
<li>Your API key will be visible in the sidebar</li>
</ol>
<p>If you are downloading or uploading large datasets from the portal, it is highly recommended that you do this from an appropriate environment, with a fast and reliable connection to the internet. Many institutions provide access to HPC nodes.</p>
<div id="getting-started-r" class="section level3">
<h3>Getting started: R</h3>
<p>You will need access to an <a href="https://www.r-project.org">R</a> installation, either on your computer or on a HPC node.</p>
<ol style="list-style-type: decimal">
<li>Launch R, and</li>
<li>Install the <a href="https://github.com/ropensci/ckanr">ckanr</a> bindings for CKAN:</li>
</ol>
<pre><code>&gt; install.packages(&quot;devtools&quot;)
&gt; library(&quot;devtools&quot;)
&gt; install_github(&quot;ropensci/ckanr&quot;)</code></pre>
</div>
<div id="getting-started-python" class="section level3">
<h3>Getting started: Python</h3>
<p>You will need access to a <a href="https://www.python.org">Python</a> installation, either on your computer or on a HPC node. <strong>Note:</strong> Python is installed by default on MacOS and most Linux systems. The examples given here are for Python version 3.7 or higher.</p>
<p>Use <code>pip</code> to install the <a href="https://github.com/ckan/ckanapi">ckanapi</a> extension and the <code>requests</code> extension:</p>
<pre><code>$ pip install ckanapi
$ pip install requests</code></pre>
</div>
<div id="getting-started-bash" class="section level3">
<h3>Getting started: Bash</h3>
<p>You will need <a href="https://curl.haxx.se/">curl</a> installed. <strong>Note:</strong> Curl is installed by default on MacOS and most Linux systems. For Windows, you can easily install it using the <a href="https://docs.microsoft.com/en-us/windows/wsl/">Windows Subsystem for Linux</a>, or <a href="https://chocolatey.org/">chocolatey</a>.</p>
<p>You will also need <a href="https://stedolan.github.io/jq/">jq</a> installed - it is including in most Linux distributions, and is available via <a href="https://brew.sh/">homebrew</a> on MacOS.</p>
</div>
</div>
<div id="searching-the-ckan-archive" class="section level2">
<h2>Searching the CKAN archive</h2>
<p>CKAN makes it possible to query datasets and resources programmatically.</p>
<p>This may be of use when:</p>
<ul>
<li>running a specific query repeatedly, or</li>
<li>building a front-end or command line tool to search for data in a particular way</li>
</ul>
<p>It is helpful to be aware of two key concepts which the Data Portal uses:</p>
<ol style="list-style-type: decimal">
<li>Packages: a package is a collection of zero or more <em>resources</em>. It is effectively a flat folder, with metadata.</li>
<li>Resources: a resource is a file, with its associated metadata, including its type and size</li>
</ol>
<p>The following sections contain example code snippets demonstrating how to search for all ‘amplicon’ data from the Australian Microbiome Framework Initiative. The Bioplatforms Australia Data Portal allows up to 50,000 search results to be returned at a time. The results returned will contain the <em>package</em> and <em>resource</em> metadata for each ‘amplicon’ dataset.</p>
<p>The list of all data types, and their associated schemas, are in the <a href="https://github.com/BioplatformsAustralia/ckanext-bpatheme/tree/master/ckanext/bpatheme">ckanext-bpatheme</a> repository on Github. Each data type has a JSON schema file in that directory. <strong>Make sure to change to delimiter from <code>_</code> (underscore) to <code>-</code> (hyphen) prior to trying a programmatic search.</strong></p>
<div id="searching-python" class="section level3">
<h3>Searching: Python</h3>
<pre class="python"><code>import ckanapi
remote = ckanapi.RemoteCKAN(&#39;https://data.bioplatforms.com&#39;, apikey=&#39;xx-xx-xx-xx-xx&#39;)
# we increase the number of rows to be returned, and we
# ask for all packages, including private packages
result = remote.action.package_search(
    q=&#39;type:amdb-genomics-amplicon&#39;,
    rows=50000,
    include_private=True)
print(&quot;{} matches found.&quot;.format(result[&#39;count&#39;]))</code></pre>
</div>
<div id="searching-r" class="section level3">
<h3>Searching: R</h3>
<pre class="r"><code>library(ckanr)
key=&quot;xx-xx-xx-xx-xx&quot;
ckanr_setup(url=&quot;https://data.bioplatforms.com&quot;, key=key)
x &lt;- package_search(q=&#39;type:amdb-genomics-amplicon&#39;, rows=50000, include_private=TRUE)
print(x$count)</code></pre>
</div>
<div id="searching-bash" class="section level3">
<h3>Searching: bash</h3>
<pre class="bash"><code>export CKAN_API_KEY=&quot;xx-xx-xx-xx-xx&quot;
curl -H &quot;Authorization: &quot;$CKAN_API_KEY&quot; &#39;https://data.bioplatforms.com/api/3/action/package_search?q=type:amdb-genomics-amplicon&amp;rows=50000&amp;include_private=true&#39;</code></pre>
</div>
</div>
<div id="downloading-data-from-the-ckan-archive" class="section level2">
<h2>Downloading data from the CKAN archive</h2>
<p>Once you have identified the packages containing the data you wish to download, it is possible to programmatically download this data. You may also wish to save the associated metadata for use in your analysis.</p>
<p>The following examples extend the search functionality above to download and save data files.</p>
<div id="downloading-data-python" class="section level3">
<h3>Downloading data: Python</h3>
<pre class="python"><code>import ckanapi
remote = ckanapi.RemoteCKAN(&#39;https://data.bioplatforms.com&#39;, apikey=&#39;xx-xx-xx-xx-xx&#39;)
# we increase the number of rows to be returned, and we
# ask for all packages, including private packages
result = remote.action.package_search(
    q=&#39;type:amdb-genomics-amplicon&#39;,
    rows=50000,
    include_private=True)
print(&quot;{} matches found.&quot;.format(result[&#39;count&#39;]))
# iterate through the resulting packages, downloading them one by one
import requests
for package in result[&#39;results&#39;]:
    for resource in package[&#39;resources&#39;]:
        url = resource[&#39;url&#39;]
        resp = requests.get(resource[&#39;url&#39;], headers={&#39;Authorization&#39;: remote.apikey})
        with open(resource[&#39;name&#39;], &#39;wu&#39;) as fd:
            fd.write(resp.content)</code></pre>
</div>
<div id="downloading-data-r" class="section level3">
<h3>Downloading data: R</h3>
<pre class="r"><code>library(ckanr)
key=&quot;xx-xx-xx-xx-xx&quot;
ckanr_setup(url=&quot;https://data.bioplatforms.com&quot;, key=key)
x &lt;- package_search(q=&#39;type:amdb-genomics-amplicon&#39;, rows=50000, include_private=TRUE)
headers &lt;- (key)
names(headers) &lt;- c(&#39;Authorization&#39;)
for (package in x$results) {
    for (resource in package$resources) {
        print(resource$url)
        download.file(resource$url, resource$name, headers=headers)
    }
}</code></pre>
</div>
<div id="downloading-data-bash" class="section level3">
<h3>Downloading data: bash</h3>
<p>We ‘pipe’ the response received from the Data Portal into <code>jq</code>, a command-line tool which allows us to easily parse JSON data.</p>
<pre class="bash"><code>export CKAN_API_KEY=&quot;xx-xx-xx-xx-xx&quot;
for URL in $( curl -H &quot;Authorization: $CKAN_API_KEY&quot; &#39;https://data.bioplatforms.com/api/3/action/package_search?q=type:amdb-genomics-amplicon&amp;rows=50000&amp;include_private=true&#39; | jq -r &#39;.result .results [] .resources [] .url&#39;); do
  # download the file using curl
  echo &quot;downloading: $URL&quot;
  curl -O -L -C - -H &quot;Authorization: $CKAN_API_KEY&quot; &quot;$URL&quot;
done</code></pre>
<p>You can also use <code>jq</code> to generate an MD5 checksum file, which can be used to verify the downloaded data.</p>
<pre class="bash"><code>export CKAN_API_KEY=&quot;xx-xx-xx-xx-xx&quot;
curl -H &quot;Authorization: $CKAN_API_KEY&quot; &#39;https://data.bioplatforms.com/api/3/action/package_search?q=type:amdb-genomics-amplicon&amp;rows=50000&amp;include_private=true&#39; | jq -r &#39;.result .results [] .resources [] | &quot;\(.md5)  \(.name)&quot;&#39; &gt; checksums.md5
# check downloaded data
md5sum -c checksums.md5</code></pre>
</div>
<div id="downloading-data-bash-hpc" class="section level3">
<h3>Downloading data: bash @ HPC</h3>
<p>If you wish to download data directly from your cloud or HPC environment, you can copy the bulk download Zip archive, described in the <a href="find_filter_download.md">find, filter, download</a> guide, to that system, extract it, and run the download (<code>download.sh</code>) from there. This means you will have access to the reliable, high-speed connection to the internet provided by that environment, as well as any attached storage resources.</p>
</div>
</div>
<div id="advanced-integrations-with-the-archive" class="section level2">
<h2>Advanced integrations with the archive</h2>
<p>Once you are set up for programmatic access, it is build more advanced integrations with the portal. As examples, you could:</p>
<ul>
<li>poll the portal periodically for new data, and automatically download it</li>
<li>only download files that you haven’t already downloaded</li>
<li>run an analysis pipeline as data becomes available</li>
<li>perform arbitrary complex filters on the data and metadata in the portal</li>
</ul>
<p>The example below (in Python) explores some of those possibilites, building on the snippets above.</p>
<pre class="python"><code>import ckanapi
import datetime
import os

remote = ckanapi.RemoteCKAN(&#39;https://data.bioplatforms.com&#39;, apikey=&#39;xx-xx-xx-xx-xx&#39;)
# we increase the number of rows to be returned, and we
# ask for all packages, including private packages
result = remote.action.package_search(
    q=&#39;type:amdb-genomics-amplicon&#39;,
    rows=50000,
    include_private=True)
print(&quot;{} matches found.&quot;.format(result[&#39;count&#39;]))
# iterate through the resulting packages, downloading them one by one
import requests
for package in result[&#39;results&#39;]:
    # filter the results based on the date they were sampled
    date_sampled = datetime.datetime.strptime(package[&#39;date_sampled&#39;], &#39;%Y-%m-%d&#39;).date()
    # if the sample is less than 90 days old, excluded it
    if (datetime.date.today() - date_sampled).days &lt; 90:
        continue
    for resource in package[&#39;resources&#39;]:
        target = resource[&#39;name&#39;]
        # do we already have the file? if so, skip the download
        if os.access(target, os.R_OK):
            continue
        url = resource[&#39;url&#39;]
        resp = requests.get(resource[&#39;url&#39;], headers={&#39;Authorization&#39;: remote.apikey})
        with open(target, &#39;wu&#39;) as fd:
            fd.write(resp.content)</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
